### ISSUE TITLE
Guice configuration errors in Play 2.4

### ISSUE THREAD

**Issue by xring** [chrisnappin/play-recaptcha-example#1]
I copy all codes in this repo to my new project with Play 2.4.2, and change the `Controllers` from `object` to `class`, something like this:

``` scala
class ExampleForm @Inject()(val messagesApi: MessagesApi) extends Controller with I18nSupport
```

and

``` scala
class Index @Inject()(val messagesApi: MessagesApi) extends Controller with I18nSupport
```

but when I run the project I got this:

> ConfigurationException: Guice configuration errors:
> 
> 1) Could not find a suitable constructor in com.nappin.play.recaptcha.RecaptchaPlugin. Classes must have either one (and only one) constructor annotated with @Inject or a zero-argument constructor that is not private.
>   at com.nappin.play.recaptcha.RecaptchaPlugin.class(RecaptchaPlugin.scala:25)
>   while locating com.nappin.play.recaptcha.RecaptchaPlugin
> 
> 1 error

![image](https://cloud.githubusercontent.com/assets/7193733/9424811/a8882f02-492c-11e5-8125-2d15b5ff3cd6.png)

and the stack trace:

```
->com.google.inject.ConfigurationException: Guice configuration errors:

1) Could not find a suitable constructor in com.nappin.play.recaptcha.RecaptchaPlugin. Classes must have either one (and only one) constructor annotated with @Inject or a zero-argument constructor that is not private.
  at com.nappin.play.recaptcha.RecaptchaPlugin.class(RecaptchaPlugin.scala:25)
  while locating com.nappin.play.recaptcha.RecaptchaPlugin

1 error
     com.google.inject.internal.InjectorImpl.getProvider(InjectorImpl.java:1042)
     com.google.inject.internal.InjectorImpl.getProvider(InjectorImpl.java:1001)
     com.google.inject.internal.InjectorImpl.getInstance(InjectorImpl.java:1051)
     play.api.inject.guice.GuiceInjector.instanceOf(GuiceInjectorBuilder.scala:321)
     play.api.Plugins$$anonfun$loadPlugins$1.apply(Plugins.scala:89)
     play.api.Plugins$$anonfun$loadPlugins$1.apply(Plugins.scala:87)
     scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:245)
     scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:245)
     scala.collection.immutable.List.foreach(List.scala:381)
     scala.collection.TraversableLike$class.map(TraversableLike.scala:245)
     scala.collection.immutable.List.map(List.scala:285)
     play.api.Plugins$.loadPlugins(Plugins.scala:87)
     play.api.Plugins$$anonfun$apply$4.apply(Plugins.scala:100)
     play.api.Plugins$$anonfun$apply$4.apply(Plugins.scala:100)
     play.api.Plugins.thePlugins$lzycompute(Plugins.scala:59)
     play.api.Plugins.thePlugins(Plugins.scala:59)
     play.api.Plugins.length(Plugins.scala:60)
     scala.collection.IndexedSeqLike$class.iterator(IndexedSeqLike.scala:90)
     play.api.Plugins.iterator(Plugins.scala:57)
     scala.collection.IterableLike$class.foreach(IterableLike.scala:72)
     play.api.Plugins.foreach(Plugins.scala:57)
     scala.collection.SeqLike$class.reverse(SeqLike.scala:273)
     play.api.Plugins.reverse(Plugins.scala:57)
     play.api.Play$$anonfun$stop$1.apply(Play.scala:108)
     play.utils.Threads$.withContextClassLoader(Threads.scala:21)
     play.api.Play$.stop(Play.scala:107)
     play.api.Play$.start(Play.scala:83)
     play.core.server.DevServerStart$$anonfun$mainDev$1$$anon$1$$anonfun$get$1$$anonfun$apply$1$$anonfun$1.apply(DevServerStart.scala:156)
     play.core.server.DevServerStart$$anonfun$mainDev$1$$anon$1$$anonfun$get$1$$anonfun$apply$1$$anonfun$1.apply(DevServerStart.scala:121)
     scala.Option.map(Option.scala:146)
     play.core.server.DevServerStart$$anonfun$mainDev$1$$anon$1$$anonfun$get$1$$anonfun$apply$1.apply(DevServerStart.scala:121)
     play.core.server.DevServerStart$$anonfun$mainDev$1$$anon$1$$anonfun$get$1$$anonfun$apply$1.apply(DevServerStart.scala:119)
     scala.util.Success.flatMap(Try.scala:230)
     play.core.server.DevServerStart$$anonfun$mainDev$1$$anon$1$$anonfun$get$1.apply(DevServerStart.scala:119)
     play.core.server.DevServerStart$$anonfun$mainDev$1$$anon$1$$anonfun$get$1.apply(DevServerStart.scala:111)
     scala.concurrent.impl.Future$PromiseCompletingRunnable.liftedTree1$1(Future.scala:24)
     scala.concurrent.impl.Future$PromiseCompletingRunnable.run(Future.scala:24)
     akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:40)
     akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:397)
     scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
     scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
     scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
     scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

So, how can I fix this problem?

**Comment by chrisnappin** [comment]
Hi xring,

  Thanks for using play-recaptcha. That looks like a generic error produced because the plugin doesn't use the Play 2.4 APIs.

I'll look into porting play-recaptcha to v2.4, and maybe converting from a Plugin to a Module. I take it you are looking (or already are) using Play 2.4 in production?

This will be a breaking (incompatible) change, so if anyone wants to use play-recaptcha with Play 2.3 they can use the current release instead (which I can branch and maintain if needed).

Cheers,

  Chris

**Comment by chrisnappin** [comment]
Hi, this has now been resolved by release 1.5, which should sync across maven central in the next couple of hours. I'll update the wiki over the next few days, but in the meantime simply copy one of the 2 example projects. This plugin has been converted to a DI module and uses the Play 2.4 APIs, other than that there is no change. Any feedback you can provide would be really useful!

Enjoy, Chris

**Comment by chrisnappin** [comment]
Hi,

 I've created v1.5 to support Play 2.4, and have just updated the wiki with instructions. Please let me know how you get on!

Cheers,

  Chris

On 22 August 2015 17:22:46 BST, xring notifications@github.com wrote:

> I copy all codes in this repo to my new project with Play 2.4.2, and
> change the `Controllers` from `object` to `class`, something like this:
> 
> ``` scala
> class ExampleForm @Inject()(val messagesApi: MessagesApi) extends
> Controller with I18nSupport
> ```
> 
> and
> 
> ``` scala
> class Index @Inject()(val messagesApi: MessagesApi) extends Controller
> with I18nSupport
> ```
> 
> but when I run the project I got this:
> 
> > ConfigurationException: Guice configuration errors:
> > 
> > 1) Could not find a suitable constructor in
> > com.nappin.play.recaptcha.RecaptchaPlugin. Classes must have either one
> > (and only one) constructor annotated with @Inject or a zero-argument
> > constructor that is not private.
> > at
> > com.nappin.play.recaptcha.RecaptchaPlugin.class(RecaptchaPlugin.scala:25)
> >  while locating com.nappin.play.recaptcha.RecaptchaPlugin
> > 
> > 1 error
> 
> ![image](https://cloud.githubusercontent.com/assets/7193733/9424811/a8882f02-492c-11e5-8125-2d15b5ff3cd6.png)
> 
> and the stack trace:
> 
> ```
> ->com.google.inject.ConfigurationException: Guice configuration errors:
> 
> 1) Could not find a suitable constructor in
> com.nappin.play.recaptcha.RecaptchaPlugin. Classes must have either one
> (and only one) constructor annotated with @Inject or a zero-argument
> constructor that is not private.
> at
> com.nappin.play.recaptcha.RecaptchaPlugin.class(RecaptchaPlugin.scala:25)
>  while locating com.nappin.play.recaptcha.RecaptchaPlugin
> 
> 1 error
> com.google.inject.internal.InjectorImpl.getProvider(InjectorImpl.java:1042)
> com.google.inject.internal.InjectorImpl.getProvider(InjectorImpl.java:1001)
> com.google.inject.internal.InjectorImpl.getInstance(InjectorImpl.java:1051)
> play.api.inject.guice.GuiceInjector.instanceOf(GuiceInjectorBuilder.scala:321)
>     play.api.Plugins$$anonfun$loadPlugins$1.apply(Plugins.scala:89)
>     play.api.Plugins$$anonfun$loadPlugins$1.apply(Plugins.scala:87)
> scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:245)
> scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:245)
>     scala.collection.immutable.List.foreach(List.scala:381)
>  scala.collection.TraversableLike$class.map(TraversableLike.scala:245)
>     scala.collection.immutable.List.map(List.scala:285)
>     play.api.Plugins$.loadPlugins(Plugins.scala:87)
>     play.api.Plugins$$anonfun$apply$4.apply(Plugins.scala:100)
>     play.api.Plugins$$anonfun$apply$4.apply(Plugins.scala:100)
>     play.api.Plugins.thePlugins$lzycompute(Plugins.scala:59)
>     play.api.Plugins.thePlugins(Plugins.scala:59)
>     play.api.Plugins.length(Plugins.scala:60)
> scala.collection.IndexedSeqLike$class.iterator(IndexedSeqLike.scala:90)
>     play.api.Plugins.iterator(Plugins.scala:57)
>     scala.collection.IterableLike$class.foreach(IterableLike.scala:72)
>     play.api.Plugins.foreach(Plugins.scala:57)
>     scala.collection.SeqLike$class.reverse(SeqLike.scala:273)
>     play.api.Plugins.reverse(Plugins.scala:57)
>     play.api.Play$$anonfun$stop$1.apply(Play.scala:108)
>     play.utils.Threads$.withContextClassLoader(Threads.scala:21)
>     play.api.Play$.stop(Play.scala:107)
>     play.api.Play$.start(Play.scala:83)
> play.core.server.DevServerStart$$anonfun$mainDev$1$$anon$1$$anonfun$get$1$$anonfun$apply$1$$anonfun$1.apply(DevServerStart.scala:156)
> play.core.server.DevServerStart$$anonfun$mainDev$1$$anon$1$$anonfun$get$1$$anonfun$apply$1$$anonfun$1.apply(DevServerStart.scala:121)
>     scala.Option.map(Option.scala:146)
> play.core.server.DevServerStart$$anonfun$mainDev$1$$anon$1$$anonfun$get$1$$anonfun$apply$1.apply(DevServerStart.scala:121)
> play.core.server.DevServerStart$$anonfun$mainDev$1$$anon$1$$anonfun$get$1$$anonfun$apply$1.apply(DevServerStart.scala:119)
>     scala.util.Success.flatMap(Try.scala:230)
> play.core.server.DevServerStart$$anonfun$mainDev$1$$anon$1$$anonfun$get$1.apply(DevServerStart.scala:119)
> play.core.server.DevServerStart$$anonfun$mainDev$1$$anon$1$$anonfun$get$1.apply(DevServerStart.scala:111)
> scala.concurrent.impl.Future$PromiseCompletingRunnable.liftedTree1$1(Future.scala:24)
> scala.concurrent.impl.Future$PromiseCompletingRunnable.run(Future.scala:24)
>     akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:40)
> akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:397)
>   scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
> scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
> scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
> scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
> ```
> 
> So, how can I fix this problem?
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/chrisnappin/play-recaptcha-example/issues/1

## 

Sent from my Android device with K-9 Mail. Please excuse my brevity.
